[MRG] Add functionality to record somatic voltage from all cells in simulation

ntolley opened the pull request:
closes #181. Starting this PR to discuss how this should be implemented. At the moment I am considering a dictionary indexed by gids, the values would be an array of the somatic voltage recorded. Alternatively it could be implemented as its own object like the `Spikes` class. 

Another point to consider is if we want to record and save voltages by default, or have it enabled when specified by the user.

jasmainak commented:
starting a PR after changing 3 lines, I think you're getting the hang of open source and open science now ;-)

umm ... the cleanest API would probably be to rename `Spikes` to something like `Cell`, then rename existing `Cell` to `CellBuilder` and then anything related to cells will be part of this `Cell` object: soma voltage, currents, spikes etc. Do you think it's doable by you or too much?

jasmainak commented:
Just to elaborate a bit. You should be able to do:

```py
>>> net = Network(...)
>>> net.cells[50].spikes
>>> net.plot_spikes()
>>> net.cells[50].vsoma
>>> plt.plot(net.cells[50].vsoma)
```

ntolley commented:
I'll give it a shot! I have a bit more free time in the coming weeks so ideally I can make decent progress.

ntolley commented:
First pass at storing voltages in the `Spikes` class. Everything seems to be working ok but I'll need to compare against the HNN recorded voltages. Here's an example the voltages from each cell class during the default simulation: 
![image](https://user-images.githubusercontent.com/55253912/94326030-5f568f00-ff56-11ea-917a-d41457ad286b.png)

jasmainak commented on hnn_core/network_builder.py:
```diff
@@ -635,6 +645,13 @@ def _record_spikes(self):
             if _PC.gid_exists(gid):
                 _PC.spike_record(gid, self._spiketimes, self._spikegids)
 
+    # setup somatic voltage recording for this node
```
looks very clean so far! Just a tiny comment about moving this comment into a one-line docstring below. It's not so kosher to have comments above functions (I know it's there in the repo but it's mostly legacy code)

go, go, go @ntolley !!

ntolley commented:
Implemented some basic slicing functionality for the spikes class (the next commit will allow for arbitrary lists/arrays) At the moment the slice operation returns smaller `Spikes` object with the content filtered by gids that match the slice. 

One consideration is that indexing gids that do not exist will return an empty list. Considering the structure of `spikes._times` and similar attributes I think this makes the most sense. However I'm wondering if a warning should be raised if slice produces a completely empty slice object.

jasmainak commented on hnn_core/network.py:
```diff
@@ -405,6 +408,40 @@ def __eq__(self, other):
                 self._gids == other._gids and
                 self._types == other._types)
 
+    def __getitem__(self, gid_item):
```
can you add a docstring above to understand what it does? Just for inspiration:

https://github.com/mne-tools/mne-python/blob/270b77768e4d2cf0f3f7ae4defd229c9f39eb2af/mne/utils/mixin.py#L93-L126

But don't get this fancy. [Keep it simple stupid](https://en.wikipedia.org/wiki/KISS_principle). MNE slicing functionality has expanded over 10 years ... it started simple.

I'm thinking that you want to be index by gid (and maybe by trial index)

jasmainak replied:
another alternative is that `Spikes` object will store only info for a single cell and then you store a list of spike objects. In that case `plot_spikes` will probably have to be a function that takes the list of spikes as input rather than a method of `spikes` class ... I kind of like this alternative better because then you can also in the future store cell geometry information in this class. Think of the pros and cons, we'll probably have to discuss IRL but so far I'm leaning towards this second alternative (depending on how easy it is to implement, don't want things to get complicated, cf KISS principle)

ntolley replied:
Storing it as a list would be less cumbersome in terms of updating information stored. As you can see a lot of code is dedicated to just reformatting the data after filtering. However it sounds like that change will require a lot of changes outside the `Spikes` class.

I'll look to characterize how wide reaching that change will be with respect to the current code. My motivation for the current change is it has zero impact on the rest of the code-base. I'll stick with this implementation while I improve how voltage recordings are passed around.

jasmainak replied:
perhaps storing `Spikes` as a list could be the next PR ... we do want to have clean data structures before making a release though.

rythorpe replied:
I really like the idea of a new `CellResponse` (or whatever we want to call it) object that stores simulation output data like spikes, soma voltage, etc. on a cell-by-cell basis. We can defs save this for a different PR though.

rythorpe commented on hnn_core/network_builder.py:
```diff
@@ -105,6 +107,10 @@ def simulation_time():
         for spike_vec in spikegids_list:
             neuron_net._all_spikegids.append(spike_vec)
 
+        for node_vsoma in vsoma_list:
+            vsoma_data = {k: v.to_python() for k, v in node_vsoma.items()}
+            neuron_net._all_vsoma.update(vsoma_data)
```
I'm still trying to wrap my head around how the various nrn and PC calls work, but it looks like `vsoma_list` contains hoc objects buried in it. Is there anyway to simplify `vsoma_list` or `neuron_net._vsoma` so that these variables are more transparent? Maybe `NetworkBuilder().vsoma` could be assigned to a special nrn hoc object that allows `_PC.py_gather()` to instantiate a python-only object?

ntolley replied:
That is correct, but you'll notice that in the immediately preceding lines that `spiketimes_list` and `spikegids_list` also contain hoc objects. The only difference is they are converted to python objects in the function:
```py
    def get_data_from_neuron(self):
        """Get copies of spike data that are pickleable"""

        from copy import deepcopy
        data = (self._all_spiketimes.to_python(),
                self._all_spikegids.to_python(),
                deepcopy(self.net.gid_dict),
                deepcopy(self._all_vsoma))
        return data
```
The main difference is where the python conversion is made. Perhaps as we restructure the spikes class we can change the variable name?

In either case I'll likely need to restructure this as I think using a dictionary is breaking MPI runs (but not joblib) for some reason. I like the suggestion of keeping `NetworkBuilder().vsoma` as a hoc object, perhaps that will help with the MPI compatability...

rythorpe replied:
> That is correct, but you'll notice that in the immediately preceding lines that `spiketimes_list` and `spikegids_list` also contain hoc objects.

Oh I see. I had thought that `_PC.py_gather()` was concatenating hoc objects and spitting out a single python object, but I see now that that is not the case.

> In either case I'll likely need to restructure this as I think using a dictionary is breaking MPI runs (but not joblib) for some reason.

I'm wondering if your issue with MPI arrises because you extract the hoc-converted-to-python data too soon.

jasmainak replied:
I didn't read carefully but I have an intuition that dictionary MPI because it's not ordered. The MPI calls depend on having things called in the right order. There is a node-specific list called `gid_list` that you should use when needed to make MPI happy.

jasmainak replied:
also you might want to rebase before debugging further

jasmainak replied:
unfortunately debugging MPI is notoriously hard. You will have to insert print statements and check where the code is failing ... also trying printing what is in `stderr`. Hope these hints hellp

ntolley commented:
Well this is no fun, the test Travis is failing on seems to run fine on my machine. In any case I restructured storage of `vsoma` to mimic `cell.current` by setting up recordings when the cell is instantiated. In the future it may be worth moving `cell.current` data to the proposed `CellResponse` class, however the previous MPI errors may reflect an issue with creating `h.Vector()` recordings in the wrong scope. 

@rythorpe @jasmainak  If you have the chance can you try running `test_compare_across_backends()` from your machine? If there is an error it will likely read "MPI simulation didn't return any data"

In the mean time I may push some variants on handling somatic voltages to try and appease Travis/MPI.

jasmainak commented:
yep I could replicate. The problem is that the you don't have `mpi4py` installed, so you're not able to replicate. Can you do:

```py
import mpi4py
```

in the ipython terminal? Also try putting a breakpoint (e.g., just add a line saying "1/0") in your test, and run it with "--pdb" flag and try doing "import mpi4py". I think it won't work.

Now the issue is the following -- we don't get a message that the test was skipped when `mpi4py` is not installed. I thought this was fixed but apparently [we didn't do the proper fix](https://github.com/jonescompneurolab/hnn-core/blob/master/hnn_core/tests/test_compare_hnn.py#L92-L100) when I looked at it. We should do [something like this](https://github.com/mne-tools/mne-python/blob/018320e72107c6cb6691772edc4f78b15cc73703/mne/viz/tests/test_3d_mpl.py#L83) so the entire test is skipped when `mpi4py` is not installed. @ntolley want to give a shot at fixing that first? Basically, [copy these lines](https://github.com/mne-tools/mne-python/blob/b62d8297f60c4b8c8f3ec44e499353b4a73efee1/mne/utils/_testing.py#L102-L115) and then rebase this pull request once that's merged into master

jasmainak commented:
also cc-ing @blakecaldwell since he's the MPI expert :-)

jasmainak commented:
By the way, just for your info. When you have something finicky that you can't replicate on your box, there is also the option of ssh-ing into the Travis build and trying there. It's a pain (because Travis box won't have a bunch of stuff installed) but you can use it as a last resort.

blakecaldwell commented:
> I thought this was fixed but apparently [we didn't do the proper fix](https://github.com/jonescompneurolab/hnn-core/blob/master/hnn_core/tests/test_compare_hnn.py#L92-L100) when I looked at it.

The way I did it preserves comparing results when the joblib backend is used. Maybe you want to split this test up like it was before https://github.com/jonescompneurolab/hnn-core/commit/e847137eec6ca75925b8d86e07b7a7a8efb29aaa#diff-c85e4dc16a1586adb223901f0de74fdf

blakecaldwell commented:
@ntolley I found a problem in how the results are being passed between `mpi_child.py` and `MPIBackend.simulate()`. This is a failure of my code to correctly pad the data. So I'll come up with a fix in new PR and you can rebase? It seems that tests past once I fix this.

@jasmainak since you have a specific plan in mind for how the tests should look, what about taking that on in a future PR?

blakecaldwell commented:
While #192 is still open, you can workaround this by 
```
diff --git a/hnn_core/mpi_child.py b/hnn_core/mpi_child.py
index 5d02ab85..ec5d34b7 100644
--- a/hnn_core/mpi_child.py
+++ b/hnn_core/mpi_child.py
@@ -92,6 +92,7 @@ def run_mpi_simulation():
                                         'base64')
 
         data_iostream.write(repickled_bytes + b"===")
+        data_iostream.write(repickled_bytes + b"====")
 
     # flush anything in stderr (still points to str_err) to stdout
     sys.stderr.flush()

```

ntolley commented:
Thanks @blakecaldwell! 

@jasmainak to finish off this PR I'll work on cleaning up indexing the `Spikes` object. Depending on how long #192 takes to merge I may start work on storing somatic currents in the same format. 

Keeping the integration timeline in mind, do you think the proposed test PR can be merged afterwards?

rythorpe commented:
> > I thought this was fixed but apparently [we didn't do the proper fix](https://github.com/jonescompneurolab/hnn-core/blob/master/hnn_core/tests/test_compare_hnn.py#L92-L100) when I looked at it.
> 
> The way I did it preserves comparing results when the joblib backend is used. Maybe you want to split this test up like it was before [e847137#diff-c85e4dc16a1586adb223901f0de74fdf](https://github.com/jonescompneurolab/hnn-core/commit/e847137eec6ca75925b8d86e07b7a7a8efb29aaa#diff-c85e4dc16a1586adb223901f0de74fdf)

Just to provide more context, the commit that combined running both backends under a single test was designed to speed up and simplify the test. The idea was that each backend needed two simulations: one for comparing to the default 'ground truth' simulation and one for comparing consistency between the other backends. It looks like we'll need to rethink this, however, given the mpi2py import issue. Maybe we can also consider shortening the default 'ground truth' simulation in order to speed up the tests that require simulations.

jasmainak commented:
we can use [pytest fixtures](https://docs.pytest.org/en/stable/fixture.html#fixtures-as-function-arguments) to avoid duplicating the same test.

jasmainak commented:
> Keeping the integration timeline in mind, do you think the proposed test PR can be merged afterwards?

I would say bring this PR to a mergeable state as soon as you can. I would also like to work on other PRs which could create merge conflicts and slow things down. Adding the currents should be easy once we have this working, no?

jasmainak commented on hnn_core/tests/test_network.py:
```diff
@@ -112,6 +112,10 @@ def test_spikes(tmpdir):
 
+    with pytest.raises(TypeError, match="indices must be integers or slices, "
+                       "not str"):
+        spikes['1']
```
need more tests!

ntolley commented:
Yeah adding currents would work nearly identically as somatic voltage. A cleaner solution might actually be to combine `aggregate_currents()` and `aggregate_voltages()` functions. This is more complicated due to how the dipole is calculated (summation occurs implicitly in the parallel call), but shouldn't be too difficult.

ntolley commented:
Will dig into the Travis error tomorrow. The error message is actually being printed now since the voltage data isn't passed through stderr, but I am unsure if it is related to how I am storing voltage data or some other issue:
![image](https://user-images.githubusercontent.com/55253912/96061264-cb932700-0e60-11eb-92a5-b066b7f35431.png)

In any case, I've flushed out `__get_item__()` function for the `Spikes` class which allows indexing by gids. @jasmainak @rythorpe as we discussed some more thought will need to go into how we want users to interact with the fully fledged `CellResponse` class, but I think this is a good start.

Once the Travis error is resolved I'd be happy to merge, please feel free to suggest changes to the gid indexing or voltage recording.

jasmainak commented:
should we remove this example: https://jonescompneurolab.github.io/hnn-core/auto_examples/plot_firing_pattern.html#sphx-glr-auto-examples-plot-firing-pattern-py

and instead use your function to plot the soma voltage?

blakecaldwell commented:
> The error message is actually being printed now since the voltage data isn't passed through stderr, but I am unsure if it is related to how I am storing voltage data or some other issue:

Oh no, I've seen this error before and it likely could relate to my last PR. I won't have time to check into it until the weekend, but if you find any reason why it's failing with this version as opposed to a previous version of this PR, that'd be helpful troubleshooting info.

ntolley commented:
@jasmainak Should I just work on modifying the file? The change really just amounts to running `simulate_dipole()` at the top.

@blakecaldwell No worries Blake, I'll start looking at your PR as I will eventually need to understand the MPI implementation. Unfortunately I am still failing to reproduce the error locally, but will post an update if I find anything.

jasmainak commented:
> @jasmainak Should I just work on modifying the file? The change really just amounts to running simulate_dipole() at the top.

Yeah, that older implementation is not very meaningful. At the time, I still didn't have a good handle on this issue. You can remove/chuck it. What I'd like to see in an example is a plot like the one you shared above.

ntolley commented:
I've updated the example so that the "same" plots are produced, but the voltage data is accessed via the `Spikes` object. The plots are rather different since `cell.plot_voltage()` actually runs a separate simulation with step current injections. @jasmainak  do you think this function is still necessary or can it be removed?

jasmainak commented:
> @jasmainak do you think this function is still necessary or can it be removed?

go ahead and remove it.

did you check if the outputs are comparable to `hnn` GUI outputs?

jasmainak commented:
Also, it would be helpful if you can do a %timeit (in ipython) or just use the `time` module to check how long does the simulation take with and without the `vsoma` being returned. Maybe on two different configurations:

1) a canonical use case for the GUI which would be a single trial parallel across neurons.
2) Multiple trials so the `vsoma` has a more significant impact

This way we can make an informed decision whether to give the user an option to turn it off.

jasmainak commented on hnn_core/network.py:
```diff
@@ -328,6 +328,11 @@ class Spikes(object):
+        Dictionary indexed by gids containing somatic voltages
+    t_vec : list
+        List of time points for samples in continuous data.
+        This includes vsoma.
```
any change you make to the object will necessarily break a round trip IO. I know we will fix this in another PR but something to be aware of

blakecaldwell commented:
@ntolley just a heads up that #196 is about to be merged and should fix the MPI Backend failures.

ntolley commented:
Ran some quick tests to see impact of storing voltage on simulation time. Below compares joblib and MPI with either 1 or 4 trials of the default parameter set. 

**somative_voltages (sec)**
**'JL_1t'**: 59.2, **'MPI_1t':** 22.7, **'JL_4t'**: 129.8, '**MPI_4t'**: 87.8

**master (sec)**
'**JL_1t'**: 58.6, **'MPI_1t'**: 31.3, **'JL_4t'**: 129.5, **'MPI_4t'**: 79.3

The impact seems fairly negligible but that may depend on RAM availability. Once the MPI fix is merged I will test with more trials (padding error only shows up with long simulations or several trials).

ntolley commented:
@jasmainak I found that one of the tests in `test_cell.py` utilized `plot_voltage()` [here](https://github.com/jonescompneurolab/hnn-core/blob/2f12489ceb8683fd674a05ccb45b93f9c06d8b34/hnn_core/tests/test_cell.py#L22). Is there another purpose to the function that I'm missing? Otherwise I'll get rid of the test as well.

jasmainak commented:
>  Is there another purpose to the function that I'm missing? Otherwise I'll get rid of the test as well.

Yep remove the function and the test.

jasmainak commented on examples/plot_firing_pattern.py:
```diff
@@ -15,8 +15,7 @@
 # Let us import hnn_core
```
should we rename the example to say `plot_somatic_voltages.py` or something like that?

ntolley replied:
I say keep the name, we can add a plot to this file for currents in the next PR.

jasmainak commented on hnn_core/tests/test_network.py:
```diff
@@ -112,6 +112,25 @@ def test_spikes(tmpdir):
+
+    with pytest.raises(TypeError, match="gids must be of dtype int, "
+                       "not float64"):
+        spikes[[0, 1, 2, 2.0]]
```
maybe also some smoke test that checks that `vsoma` > -65 or whatever makes sense?

ntolley replied:
Really good idea! Most comprehensive test would be to check that voltages above a certain threshold align with spike times. Perhaps that can be added to a future PR, with just the baseline check for now.

Important to note that vsoma is only initialized after a simulation run. This could be added to test_dipole() at the risk of poor file naming. I'd prefer not to run an extra sim in this file as it would slow down the test.

blakecaldwell replied:
An option would be to run a reduced (25 ms) simulation here. However, in the interest of comparing all outputs for similarity to HNN (@rythope), `test_compare_across_backends` would be a good place to check the vsoma time-series, similar to dipole.
https://github.com/jonescompneurolab/hnn-core/blob/master/hnn_core/tests/test_parallel_backends.py#L62

jasmainak replied:
@ntolley I think you should use `pytest` fixtures for avoiding multiple simulations. And @blakecaldwell 's idea is good too. Let me know if you have trouble figuring it out. I can point you to some example code.

jasmainak replied:
Also okay to add the second test in another PR if you prefer. We need to keep productive contributors motivated !

jasmainak commented:
You will need to rebase against master now to get @blakecaldwell 's new changes

blakecaldwell commented on examples/plot_firing_pattern.py:
```diff
@@ -30,26 +29,33 @@
-        ax.set_title('%s (gid=%d)' % (cells[gid].name, gid))
+# # Let's do this for the rest of the cell types
+fig, axes = plt.subplots(1, 2, sharey=True, figsize=(8, 4))
+for gid, ax in zip([35, 170], axes):
```
curious where gids 35 and 170 came from @jasmainak ? if they are of a certain type, would it be useful to add that in a comment?

jasmainak replied:
sure, or perhaps use `gid_dict['L2_Pyrmaidal'][0]` or some such thing

blakecaldwell replied:
> sure, or perhaps use `gid_dict['L2_Pyrmaidal'][0]` or some such thing

This is excellent. Almost self-documenting code.

ntolley replied:
> Almost self-documenting code.

Good evidence that hnn-core is improving! Also added this modification, looks like the numbers were chosen because they are the boundaries of the gid ranges which define the cell classes.

blakecaldwell commented on hnn_core/network.py:
```diff
@@ -328,6 +328,11 @@ class Spikes(object):
         Each gid corresponds to a type via Network::gid_dict.
+    vsoma : dict
+        Dictionary indexed by gids containing somatic voltages
+    t_vec : list
+        List of time points for samples in continuous data.
+        This includes vsoma.
```
I know `t_vec` is extensively used in HNN GUI, but the only place I can see it used here is in the example. Other code in this PR that creates the NEURON object, sets up the recording, and copies after the end of simulation might be unnecessary if the example (plotting) recreated t_vec as a numpy array from 0 to `t_stop` with `dt`.
https://github.com/jonescompneurolab/hnn-core/blob/7f142513bd7d42e6dfcd660126551d3962fabb07/examples/plot_simulate_gamma.py#L57

ntolley replied:
I understand your point, however this relies on simulation time always increasing in increments of `dt`. We we sure this is always the case? Especially during parallel runs.

blakecaldwell replied:
Yes, we never use the variable step integration method. I imagine this choice was made early on. More recently, Michael Hines said we might compare the two, so I tried the variable step method and it slows down the simulation. Using ParallelContext does not affect this.

ntolley replied:
Good to know. In that case what do you think of removing the t_vec recording that is passed to the `Dipole` object as well? If it's equivalent to `np.arange()` we may as well reduce the amount of data being passed around.

blakecaldwell replied:
In principle, I agree with using `np.arrange()` instead of, in each case, creating a `t_vec` hoc object, copying with `.to_python()`, and then passing around. This will become a rather common task in `Dipole` and `Network`, so I wonder if this justifies a utility function with `tstop` and `dt` parameters? I know it's just a one-liner, but it would help to keep this process standard (i.e. t[-1]=170.0 vs. only t[-1]=169.75). I don't know where to put it though... @jasmainak?

jasmainak replied:
Ideally, the code should be structured so that `times` is "inherited" from other objects without recreating it. For example, if the `Network` object creates it once, then the `Dipole` and `Spikes` object should copy that. That way you make sure there are no bugs ... even if you use variable time etc. It's what we do in MNE, for example when [creating evoked from epochs](https://github.com/mne-tools/mne-python/blob/master/mne/epochs.py#L1038)

blakecaldwell commented on hnn_core/basket.py:
```diff
@@ -69,6 +69,7 @@ def __init__(self, gid=-1, pos=-1):
         self._synapse_create()
         self._biophysics()
         self.sect_loc = dict(proximal=['soma'], distal=['soma'])
+        self.record_voltage_soma()
```
I'm a little worried that this isn't the option of turning this off (the default in HNN GUI). Would it be possible to use a new parameter `record_vsoma` in lieu of the HNN  GUI parameter `save_vsoma`?

blakecaldwell replied:
This is all that's needed to add a parameter:
https://github.com/jonescompneurolab/hnn-core/pull/184/files#diff-7fab1444eb40a87689cc4afa7c7aa15c80dc4a18a2179d3e231185de704c3bcfR138

jasmainak replied:
I think a parameter to the `__init__` of the cell class should work

ntolley replied:
Given all the issues that are cropping up with the increased data flow through stderr, I agree this will be important to disable by default (despite the negligible impact on simulation time). 

I'll combine both suggestions. `record_vsoma` will be set to `False` in the default params. `network_builder.py` will pass this value to the `__init__` of `cell.py` and a logical statement will check and return an empty list.

ntolley replied:
I've implemented it such that `params['vsoma']` is pulled out when the cells are built in `network_builder.py`. Setting `record_vsoma` = 0 makes `cell.rec_v` creates an empty `h.Vector()` in place of a recording.

ntolley commented:
@blakecaldwell It looks like your data length check is getting raised, perhaps this is a problem with my implementation? 
![image](https://user-images.githubusercontent.com/55253912/96733652-b3069d80-1387-11eb-90a0-b63e298c3d95.png)

blakecaldwell commented:
> @blakecaldwell It looks like your data length check is getting raised, perhaps this is a problem with my implementation?
> ![image](https://user-images.githubusercontent.com/55253912/96733652-b3069d80-1387-11eb-90a0-b63e298c3d95.png)

It'll take some debugging to figure this out. Would you be willing to try some things? The spikes object now has somatic voltages. Can you tell if the inclusion of that within `net.spikes` is causing the error?

ntolley commented:
@blakecaldwell It looks like the point of failure is `spikedata = neuron_net.get_data_from_neuron()` which is called from `_clone_and_simulate()`.

Commenting out the lines which converts `vsoma` from a hoc to python object seems to resolve the error.

blakecaldwell commented:
> Commenting out the lines which converts `vsoma` from a hoc to python object seems to resolve the error.

Thanks. So it seems when `vsoma_py` is empty, that it succeeds. I've been able to get it to work on Travis spuriously, but haven't found the exact issue yet.

jasmainak commented on examples/plot_firing_pattern.py:
```diff
@@ -30,26 +29,33 @@
+
+# The cell IDs (gids) are stored in the network object as a dictionary
+gid_dict = net.gid_dict
+print(net.gid_dict)
```
not sure if you intended to create a new cell here with an output but if you want to do that, you can simply add:

###############################################################################

jasmainak commented on hnn_core/pyramidal.py:
```diff
@@ -291,6 +291,7 @@ def __init__(self, gid=-1, pos=-1, p={}):
 
         # run record_current_soma(), defined in Cell()
         self.record_current_soma()
+        self.record_voltage_soma()
```
why is this added separately for all the cell classes instead of the base class?

ntolley replied:
Mainly to keep consistency with how `record_current_soma()` is implemented. Given the similarity in their implementation I think it's less confusing if the code is grouped like this.

I can change it now, but I'd prefer to hold off until the "record current" PR since that will likely restructure the existing template for storing continuous data.

jasmainak replied:
ok, sounds good. Let's keep tab of things you are postponing to the next PR. So we don't forget

jasmainak commented on examples/plot_firing_pattern.py:
```diff
@@ -30,26 +29,33 @@
+plt.title('%s (gid=%d)' % (net.gid_to_type(gid), gid))
+plt.xlabel('Time (ms)')
+plt.ylabel('Voltage (mV)')
+plt.show()
```
do you also want to plot the spikes in another subplot for one cell? We already have the function. Got the idea from your comment in the test :)

ntolley replied:
May as well do the raster plot since it is the "firing pattern" example.

blakecaldwell commented:
@ntolley after lost of time searching, this is most likely the culprit.
https://github.com/jonescompneurolab/hnn-core/pull/198/commits/2e425c53ad60fe223ede18050a52d047c2202447

My PR #198 ended up becoming rather large, so it may take longer to merge. Perhaps you can `git cherry-pick` that commit to move forward.

blakecaldwell commented:
> @ntolley after lost of time searching, this is most likely the culprit.
> [2e425c5](https://github.com/jonescompneurolab/hnn-core/commit/2e425c53ad60fe223ede18050a52d047c2202447)
> 
> My PR #198 ended up becoming rather large, so it may take longer to merge. Perhaps you can `git cherry-pick` that commit to move forward.

I take this back. The whole PR is needed.

jasmainak commented on hnn_core/network.py:
```diff
@@ -160,9 +160,10 @@ def __init__(self, params):
-                                 self.params['dt']).size + 1
+        # or not cells of relevant type actually
+        self.t_vec = np.arange(0., params['tstop'] + params['dt'],
+                               params['dt'])
```
can we call `t_vec` to be `times`? It's more descriptive. Also is `n_times` attribute really needed? If not, let's remove it. Also update the docstring!

ntolley replied:
Renaming to `times` should be fine as far as I can tell. More consistent too since `t_vec` is passed to the `times` attribute of `Dipole`.

`n_times` is never used explicitly, but it is referenced in the doc string of every example, as well as the `Dipole` class. Obviously someone thought it was important at some point, but it's pretty trivial remove if you still want to.

ntolley replied:
I spoke too soon. `times` is an attribute of the `Spikes` class. Perhaps we can rename that to `spiketimes`?

jasmainak replied:
yep, let's rename that one to `spike_times` and we can remove `n_times` if it's not used anywhere

rythorpe replied:
I think we're verging into a problem caused by conflating two classes that shouldn't be combined. `Spikes` was originally designed to represent instances of spikes. As such, it has corresponding attributes of `Spikes.times`, `Spikes.gids`, and `Spikes.types`. In order to provide a clear and consistent API, I think there are a few options we should consider:

1. Separate the `Spikes` class from the `CellResponse` class
2. If we want to rename `Spikes.times` to `Spikes.spiketimes`, I think we should rename all of the other spike-related attributes to be consistent (e.g., also rename the `gids` attribute to `spikegids`)
3. Rename `Spikes.times` to something like `Spikes.time_trajectory` to indicate that the time values specified in this attribute include all time steps rather than just the ones that pertain to spikes

jasmainak replied:
Either 2. or 3. is fine for me. Maybe slight preference for 2. since `times` is used elsewhere in `Dipole` object etc. with the same meaning, so you'd have to change it there too if you went with 3.

ntolley replied:
I agree, my preference is for 2 as well. Given how closely coupled spikes, currents, and voltages are I think it is best to keep them grouped and rename `Spikes` to `CellResponse` either in this PR, or in #199.

I've been putting off renaming since just adding voltages doesn't seem enough to justify it yet. Given the confusion @rythorpe is bringing up, do you think it should be done in this PR?

rythorpe replied:
Maybe let's save it for #199 so that we can move forward with this PR.

ntolley commented:
I've addressed the comments/suggestions raised above. I'm pretty satisfied with the state of this PR and have changed the status to [MRG]. What remains code review and fixing lingering travis errors. 

A bonus of adding the ability to turn off the voltage recording is we don't need to wait for the MPI fix. `test_parallel_backends.py` passes just fine due to the recording being disabled by default.

Unfortunately a new error has been raised that is rather cryptic:
![image](https://user-images.githubusercontent.com/55253912/97787707-4d39c300-1b8a-11eb-8ac2-c2a19e9d6a0a.png)

@jasmainak @rythorpe have either of you encountered this before?

jasmainak commented on examples/plot_firing_pattern.py:
```diff
@@ -15,41 +15,64 @@
+# Then we read the parameters file and enable somatic voltage recording
 params_fname = op.join(hnn_core_root, 'param', 'default.json')
 params = read_params(params_fname)
+params['record_vsoma'] = 1
```
this is unpythonic and makes documentation of parameters hard. I want to discourage this pattern in the command line. Can we instead add this parameter to one of the functions like `simulate_dipole` ?

jasmainak commented on hnn_core/basket.py:
```diff
@@ -64,25 +64,27 @@ def _synapse_create(self):
     """Class for layer 2 basket cells."""
 
-    def __init__(self, gid=-1, pos=-1):
+    def __init__(self, gid=-1, pos=-1, record_vsoma=0):
```
don't use `int` for boolean ... also you need to document the parameter

jasmainak commented on hnn_core/cell.py:
```diff
@@ -239,6 +242,13 @@ def record_current_soma(self):
+        if self.record_vsoma:
+            self.rec_v = h.Vector().record(self.soma(0.5)._ref_v)
+        else:
+            self.rec_v = h.Vector()
```
we're not super consistent but I believe we should document all the attributes too.

Also, it would be preferable to have `self.rec_v = h.Vector()` in the init so that here you can simply do:

```py
self.rec_v.record(....)
```

jasmainak commented on hnn_core/network.py:
```diff
@@ -301,33 +303,38 @@ class Spikes(object):
     Parameters
     ----------
-    times : list (n_trials,) of list (n_spikes,) of float, shape | None
+    spiketimes : list (n_trials,) of list (n_spikes,) of float, shape | None
```
nitpick but `spike_times` is preferred 

if you don't put the underscore it can become confusing where the second word starts. Not a problem here but generally a good practice.

jasmainak commented on hnn_core/cell.py:
```diff
@@ -239,6 +242,13 @@ def record_current_soma(self):
 
+    def record_voltage_soma(self):
+        """Record current at soma."""
+        if self.record_vsoma:
```
is it necessary to expose this attribute to the user?

rythorpe replied:
@jasmainak are you critiquing the scope of this attribute or that the `record_vsoma` arg set in the constructor is set as an attribute rather just directly passed to `_Cell.record_voltage_soma()`?

jasmainak replied:
I was thinking it could be a private attribute. But on second thoughts, I think it can stay as it is ... at least for now

jasmainak reviewed (commented):
Mostly nitpicks. I'm fine with merging this before the MPI issue is fixed as it seems a bit orthogonal. However it's likely the two PRs will be ready at the same time :)

jasmainak commented:
> > @jasmainak @rythorpe have either of you encountered this before?

just to confirm it's not a problem building the mod files on Travis are you able to reproduce this error locally?

ntolley commented:
Unfortunately I am not. Is my PR somehow impacting the mod file build or is it unrelated?

jasmainak commented:
okay then your best bet is to ssh into Travis: https://docs.travis-ci.com/user/running-build-in-debug-mode/. You're going to go ninja-level fixing this bug :)

rythorpe commented:
> I've addressed the comments/suggestions raised above. I'm pretty satisfied with the state of this PR and have changed the status to [MRG]. What remains code review and fixing lingering travis errors.
> 
> A bonus of adding the ability to turn off the voltage recording is we don't need to wait for the MPI fix. `test_parallel_backends.py` passes just fine due to the recording being disabled by default.
> 
> Unfortunately a new error has been raised that is rather cryptic:
> ![image](https://user-images.githubusercontent.com/55253912/97787707-4d39c300-1b8a-11eb-8ac2-c2a19e9d6a0a.png)
> 
> @jasmainak @rythorpe have either of you encountered this before?

I was able to reproduce this error on my machine (ubuntu), however, I'm a bit baffled by how the custom mechanisms get loaded when running tests outside of `NetworkBuilder` on the master branch. I feel like I need to understand this better before I can begin troubleshooting. @jasmainak do you understand how this works?

jasmainak commented:
you load custom mechanisms using this line: https://github.com/jonescompneurolab/hnn-core/blob/a1cc05d9146ebab3bf9623cb9611b722fee55c63/hnn_core/network_builder.py#L159.

By the way, I encountered the same error today while working on #201 . In my case, it was a bit of red herring. The actual problem was that some Neuron variable was being deleted outside the scope of the class (remember our old issues?), so adding it as an attribute of the class solved the problem.

rythorpe commented:
> you load custom mechanisms using this line:
> 
> https://github.com/jonescompneurolab/hnn-core/blob/a1cc05d9146ebab3bf9623cb9611b722fee55c63/hnn_core/network_builder.py#L159
> .
> 

Sure, but how does `test_cell.test_artificial_cell()` call `network_builder.load_custom_mechanisms()`?

```python
def test_artificial_cell():
    """Test artificial cell object."""
    event_times = [1, 2, 3]
    threshold = 0.0
    artificial_cell = _ArtificialCell(event_times, threshold)
    assert artificial_cell.nrn_eventvec.to_python() == event_times
    # the h.VecStim() object defined in vecevent.mod should contain a 'play()'
    # method
    assert hasattr(artificial_cell.nrn_vecstim, 'play')
    # the h.Netcon() instance should reference the h.VecStim() instance
    assert artificial_cell.nrn_netcon.pre() == artificial_cell.nrn_vecstim
    assert artificial_cell.nrn_netcon.threshold == threshold
```
The only reason I care is because if I start a new ipython kernel, import the appropriate modules, and try instantiating an `_ArtificialCell()` object out of the blue, an error is produced because the custom mechanisms weren't loaded.

> By the way, I encountered the same error today while working on #201 . In my case, it was a bit of red herring. The actual problem was that some Neuron variable was being deleted outside the scope of the class (remember our old issues?), so adding it as an attribute of the class solved the problem.

I was wondering about that. The trick is going to be tracing the creation/destruction of references to nrn objects....

jasmainak commented:
> Sure, but how does test_cell.test_artificial_cell() call network_builder.load_custom_mechanisms()?

humm indeed, this looks like something that needs to be fixed. Maybe call `load_custom_mechanisms` at the start of the test?

rythorpe commented on hnn_core/tests/test_cell.py:
```diff
@@ -13,14 +13,6 @@
-    params = read_params(params_fname)
-
-    net = Network(params)
-    with NetworkBuilder(net) as neuron_net:
-        neuron_net.cells[0].plot_voltage()
```
@ntolley I think this is the change that caused Travis to fail. Since `load_custom_mechanisms()` is currently only called within a `NetworkBuilder` instance, it needs to be called separately in order to successfully instantiate an `_ArtificialCell` object. If you import `load_custom_mechanisms` from `hnn_core.network_builder` and run it at the beginning of test_cell.py, pytest runs perfectly.

ntolley replied:
Really appreciate the help with this, managed to reproduce locally and this immediately resolved the error.

Should we consider an alternative place the `load_custom_mechanisms()` call? I'm not sure how frequently this sort issue would turn up outside of writing tests, but it seems like a non-intuitive bug to trace.

jasmainak replied:
maybe in the init.py file, not sure ... I'd suggest trying this in a separate PR though. Make the tests happy here so we can move on

jasmainak commented on hnn_core/cell.py:
```diff
@@ -62,19 +62,22 @@ class _Cell(ABC):
     """
 
-    def __init__(self, gid, soma_props):
+    def __init__(self, gid, soma_props, record_vsoma=0):
```
I think what you need to do is have `record_vsoma` as an argument to `simulate_dipole` and then pass it down the functions until you instantiate it. Then you don't need this parameter to the `Cell` class. You can do something like:

```py
cell = L5Pyr()
if record_vsoma: 
    cell.record_voltage_soma()
```

It's a bit odd to have it for individual cells since we're going to enable or disable it for all cells. I presume things will break if you set `record_vsoma=True` for some cells and `False` for others.

ntolley replied:
I don't believe it would break, the `vsoma` would just return an empty list for gids with `record_vsoma=False`. 

I agree with passing this as an argument to `simulate_dipole`, but it could be beneficial to have an option to select specific cells to record from in the interest of simulation time. I'll go with your recommendation for now to get this PR moving, and consider the pros/cons of enabling recordings for all cells in the current recording PR.

ntolley commented:
Never thought I'd see green on this PR. Everything seems to be functioning well on my end, let me know if there's any remaining changes you guys think are necessary. If they're not pressing they can be added to-do list for the current recording PR.

jasmainak commented on hnn_core/cell.py:
```diff
@@ -67,6 +67,9 @@ class _Cell(ABC):
     ----------
     pos : list of length 3
         The position of the cell.
+    rec_v : h.Vector()
```
why is it called `rec_v` when the parameter is called `record_vsoma` ? I think `vsoma` makes more sense as a name?

ntolley replied:
This follows the naming scheme above with `rec_i` for currents. I don't think it should be called `vsoma` because this is the only place the voltage is stored as a hoc object. `vsoma` is a python dictionary everywhere else.

jasmainak commented on examples/plot_firing_pattern.py:
```diff
@@ -15,13 +15,12 @@
 
 ###############################################################################
-# Then we read the parameters file
+# Then we read the parameters file and enable somatic voltage recording
```
shouldn't this go to the blurb below now that we do it through `simulate_dipole`?

jasmainak commented on hnn_core/network.py:
```diff
@@ -363,45 +370,116 @@ def __init__(self, times=None, gids=None, types=None):
+                self._spike_types == other._spike_types)
+
+    def __getitem__(self, gid_item):
+        """Returns a Spikes object with a copied subset filtered by gid.
```
I'm guessing we'll revisit this soon in a future PR? So I won't look carefully into this ...

ntolley replied:
That's the plan, the current implementation just exists to test the functionality without impacting any existing code.

jasmainak commented on hnn_core/tests/test_dipole.py:
```diff
@@ -53,3 +53,5 @@ def test_num_trials():
     with pytest.raises(ValueError, match="Invalid number of simulations: 0"):
         simulate_dipole(net, n_trials=0)
+    with pytest.raises(TypeError, match="record_vsoma must be bool, got int"):
+        simulate_dipole(net, n_trials=1, record_vsoma=0)
```
perhaps also a check that `net.vsoma` is `None` or the attribute is missing when `record_vsoma=False` ?

ntolley replied:
So there's two paths with this. The first would be enabling voltage recording in the `simulate_dipole()` call of `test_parallel_backends.py`. The second would be running a separate simulation for voltage recording. My preference is for the first as this keeps the test time short.

In either case enabling voltage recordings during the tests will likely reintroduce the unresolved MPI errors. It's not ideal to merge this PR without full tests, but their introduction will require #198 to be merged.

jasmainak replied:
> My preference is for the first as this keeps the test time short.

you can still use a tiny network with a small `tstop` so it runs super-fast. We don't check the voltage traces, just that the dimensions and object behave as we want.

jasmainak replied:
> enabling voltage recordings during the tests will likely reintroduce the unresolved MPI errors.

just do it for the joblibs backend. We can improve it in later PRs but having a placeholder test will help.

jasmainak commented on hnn_core/tests/test_network.py:
```diff
@@ -85,32 +85,57 @@ def test_spikes(tmpdir):
+
+    with pytest.raises(TypeError, match="gids must be of dtype int, "
+                       "not float64"):
+        spikes[[0, 1, 2, 2.0]]
```
are there any tests that check the dimensions of `vsoma` and if they match the number of trials and simulation length etc?

jasmainak commented:
@ntolley looks promising!

I left some minor comments which can be addressed easily I believe. Also when the CIs turn green, can you ensure that all 3 CIs are green. For some reason Travis didn't run on your last push. Not sure why, it can sometimes happen. Just try pushing a new commit and it might just work. Hope to merge it in a couple of days :) There is more work to do, particularly with respect to standardizing the different kinds of recordings (currents, dipoles etc) but we can keep it for future PRs! Thanks for all the hard work so far.

jasmainak commented:
You will need to rebase @ntolley since the MPI PR is now merged

ntolley commented:
Finished adding tests for the dimensions of that `vsoma` data. The MPI fix seems to be running flawlessly, thanks for all the hard work @blakecaldwell!

jasmainak commented on hnn_core/tests/test_dipole.py:
```diff
@@ -45,11 +45,23 @@ def test_dipole(tmpdir):
+
+    simulate_dipole(net, n_trials=2, record_vsoma=True)
+    assert len(net.spikes.vsoma) == 2
+    assert len(net.spikes.vsoma[0]) == 24
```
I'm going to merge this, but could you clarify where 24 comes from?

ntolley replied:
It's just the length of the recording. I figure hard coding is preferable in this situation since the length of all the recording vectors is derived from `spikes.times`. For example, it'd be useful to know if a future commit somehow changes the length of the simulation.

jasmainak replied:
I see ... you could have also used `tstop` and `dt` from `params` and recomputed the length, I guess. Anyway, no strong feelings

jasmainak commented:
Thanks a bunch @ntolley and for your patience ! 🍻 🍺 

let's keep the ball rolling in #199

blakecaldwell commented on hnn_core/dipole.py:
```diff
@@ -48,6 +50,12 @@ def simulate_dipole(net, n_trials=None):
     if n_trials < 1:
         raise ValueError("Invalid number of simulations: %d" % n_trials)
 
+    if record_vsoma is not None and isinstance(record_vsoma, bool):
+        net.params['record_vsoma'] = record_vsoma
+    else:
+        raise TypeError("record_vsoma must be bool, got %s"
+                        % type(record_vsoma).__name__)
+
```
Hi @ntolley @jasmainak! I had a followup on this PR. Enforcing a bool here causes an issue with how the HNN GUI displays parameter values. Normally, the user enters 0 or 1, which gets saved to the params dict. Here, it changes that value and the GUI shows "False", which is not suitable for a text box. I could change the GUI to automatically reconvert to 0 or 1 pretty easily, but this seems to be the first parameter hnn-core enforces. Is that the convention you are aiming for?

BTW, I realize a better GUI solution would be checkboxes but am kind of hoping hnn-core will allow 0 or 1 and I can leave the code as-is.

ntolley replied:
I'd defer to @jasmainak's opinion for this one, but I but I believe this has come up in the past and the decision was to avoids `int`s for boolean values.

jasmainak replied:
yeah, we're aiming to have this to be boolean in `hnn-core`. Do you mind type-casting it in HNN for now? We can convert to checkboxes in the future!

blakecaldwell replied:
sure, thanks for the clarification
